source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -933,6 +995,18 @@ struct FunctionParameterSpecializationContext
             newBodyInst->insertBefore(newFirstOrdinary);
         }
 
+        // We need to handle a corner case where the new argument of
+        // the callee of this specialized function could be a use of
+        // NonUniformResourceIndex(), in such case, any indexing operation
+        // on the global buffer by using this new argument should be
+        // decorated with NonUniformDecoration. However, inside the new
+        // specialized function, we don't have that information anymore.
+        // Therefore, we will need to scan the new argument list to find out
+        // this case, and insert the NonUniformResourceIndex() instruction
+        // on the corresponding parameter of the new specialized function.
+        maybeInsertNonUniformResourceIndex(newFunc, funcInfo, callInfo);
```

Comment:
If we call the function once with no uniform and once without, what do we generate? It seems that non-uniform-ness also need to be a part of the specialization key.

Reply:
Right, this is a real concern.
So it looks like we really need to change the key to make it including more modifiers.

Reply:
Today I revisited this problem, and I think we're fine in this case.
Because I noticed that the `index` parameter has already encode the uniformity info, so if there is another call site with  no uniform, the existing logic will specialize a new function for this call.

I actually add a new test case to cover this situation.

And this also replies your question here: https://github.com/shader-slang/slang/pull/4769#discussion_r1702221574

Reply:
I am not getting why you say "the existing logic will specialiize a new function for this call`. It seems that according to `getCallInfoForArg`, the `index` argument is not added to the specialization key, so if you have two callsites on the same resource array, one with nonuniform index and the other with uniform index, we will reuse the same specialization instead of producing different specializations.

Your test case is producing two specializations because the two calls feeds in different buffer arrays.

Reply:
In `getCallInfoForArg`, we add `oldIndex` into the call info key:
`ioInfo.key.vals.add(oldIndex);`

If one callsite use nonuniform index, then this `oldIndex` will be `kIROp_NonUniformResourceIndex`, and if there is another callsite using uniform index, it will be other type of inst instead of `kIROp_NonUniformResourceIndex`.

So we will have two different keys for those two callsites.

I can change my test to use the same buffer arrays, and it will still generate two different functions for those two callsites.

Reply:
Please look at my updated test, I already make two callsites on the same buffer, one with nonuniform index, and one without.

Reply:
I saw that change now, but as mentioned in the thread above, I think adding the index inst directly into the key is too much. We should only add the type of the index and the uniformity to the key.

Reply:
Let's close this thread as duplication below.